dashboard/modules/job/job_head.py

```diff
@@ -128,7 +132,66 @@ class JobHead(dashboard_utils.DashboardHeadModule):
     def __init__(self, dashboard_head):
         super().__init__(dashboard_head)
         self._dashboard_head = dashboard_head
-        self._job_manager = None
+        self._job_info_client = None
+        # this is a queue of JobAgentSubmissionClient
+        self._agents = OrderedDict()
+
+        self._agents_pool = dict()
+
+    async def choose_agent(self) -> Optional[JobAgentSubmissionClient]:
```

Comment:
This function seems wired. Do you think we should just do:
1. have a list of candidates and randomly pick one
2. another place check the liveness of this candidates and maintain the lists


If 1 returned a dead one, make it as dead and retry 1).

Reply:
It seems like this just changed from FIFO to random, maybe we just randomly choose one from available agents?

Reply:
Random sounds better to me because then we don't have to maintain any state and it seems simpler and less error-prone.  I'm not too familiar with the best practices here though.

Reply:
I refactored once, Strategy:
    1. if the number of `self._agents` has reached `CANDIDATE_AGENT_NUMBER`, randomly select one agent from `self._agents`.
    2. if not, randomly select one agent from all available agents, it is possible that the selected one already exists in `self._agents`.

Reply:
> Thanks for the update--I think I'm still not sure I understand the implementation. Won't `_agents` and `agents_pool` always be the same? It looks like they're both always popped from at the same time, and added to at the same time.

ooh, I think you are right, `_agent_pool` is redundant, I remove it.